desc/compute/_laplace.py

```diff
@@ -0,0 +1,926 @@
+    interpolator = kwargs.get("eval_interpolator", data.get("interpolator", None))
+    sign = 1 - 2 * int("exterior" in kwargs.get("problem", ""))
+
+    if kwargs["on_boundary"]:
+        RpZ_data["∇φ"] = (
+            sign
+            * 2
+            * singular_integral(
+                RpZ_data,
+                data,
+                interpolator,
+                _kernel_BS_plus_grad_S,
+                chunk_size=chunk_size,
+            )
+        )
+    else:
+        grid = transforms["grid"]
+        eval_data, source_data = _prune_data(
+            RpZ_data, None, data, grid, _kernel_BS_plus_grad_S
+        )
+        RpZ_data["∇φ"] = sign * _nonsingular_part(
+            eval_data,
+            None,
+            source_data,
+            grid,
+            st=jnp.nan,
+            sz=jnp.nan,
+            kernel=_kernel_BS_plus_grad_S,
+            chunk_size=chunk_size,
+        )
+    return RpZ_data
+
+
+@register_compute_fun(
+    name="B",
+    label="B",
+    units="T",
+    units_long="Tesla",
+    description="Magnetic field",
+    dim=3,
+    coordinates="RpZ",
+    params=[],
+    transforms={},
+    profiles=[],
+    data=["∇φ", "B0"],
+    parameterization="desc.magnetic_fields._laplace.SourceFreeField",
+)
+def _total_B(params, transforms, profiles, data, RpZ_data, **kwargs):
+    RpZ_data["B"] = RpZ_data["∇φ"] + RpZ_data["B0"]
+    return RpZ_data
+
+
+@register_compute_fun(
+    name="B0*n",
+    label="B_0 \\cdot n_{\\rho}",
+    units="T",
+    units_long="Tesla",
+    description="Auxillary field dotted into flux surface normal",
+    dim=1,
+    coordinates="tz",
+    params=[],
+    transforms={"grid": []},
+    profiles=[],
+    data=[],
+    parameterization="desc.magnetic_fields._laplace.SourceFreeField",
+    B0="_MagneticField : Field object to compute with.",
+    surface="Surface : Surface to compute on.",
+    chunk_size=_doc["chunk_size"],
+)
+def _B0_dot_n(params, transforms, profiles, data, **kwargs):
+    grid = transforms["grid"]
+    data["B0*n"] = kwargs["B0"].compute_Bnormal(
+        kwargs["surface"],
+        eval_grid=grid,
+        source_grid=grid,
+        vc_source_grid=grid,
+        chunk_size=kwargs.get("chunk_size", None),
+    )[0]
+    return data
+
+
+@register_compute_fun(
+    name="B0",
+    label="B0",
+    units="T",
+    units_long="Tesla",
+    description="Auxillary field",
+    dim=3,
+    coordinates="RpZ",
+    params=[],
+    transforms={"grid": []},
+    profiles=[],
+    data=[],
+    parameterization="desc.magnetic_fields._laplace.SourceFreeField",
+    chunk_size=_doc["chunk_size"],
+    B0="_MagneticField : Field object to compute with.",
+    public=False,
+)
+def _B0_field(params, transforms, profiles, data, RpZ_data, **kwargs):
+    coords = jnp.column_stack([RpZ_data["R"], RpZ_data["phi"], RpZ_data["Z"]])
+    RpZ_data["B0"] = kwargs["B0"].compute_magnetic_field(
+        coords=coords,
+        source_grid=transforms["grid"],
```

Comment:
I am assuming all magnetic field objects copy the source grid to each field period and integrate over the full domain [0, 2pi).